tensorwaves/data/generate.py

```diff
@@ -0,0 +1,109 @@
+"""Tools to facilitate data sample generation."""
+
+import logging
+
+import numpy as np
+
+from progress.bar import Bar
+
+from tensorwaves.interfaces import (
+    Function,
+    Kinematics,
+    PhaseSpaceGenerator,
+    UniformRealNumberGenerator,
+)
+
+
+def _generate_data_bunch(
+    bunch_size: int,
+    phsp_generator: PhaseSpaceGenerator,
+    random_generator: UniformRealNumberGenerator,
+    intensity: Function,
+    kinematics: Kinematics,
+) -> np.ndarray:
+    phsp_sample, weights = phsp_generator.generate(
+        bunch_size, random_generator
+    )
+    dataset = kinematics.convert(phsp_sample)
+    intensities = intensity(dataset)
+    maxvalue = np.max(intensities)
+
+    uniform_randoms = random_generator(bunch_size, max_value=maxvalue)
+
+    phsp_sample = phsp_sample.transpose(1, 0, 2)
+
+    return (phsp_sample[weights * intensities > uniform_randoms], maxvalue)
+
+
+def generate_data(
+    size: int,
+    phsp_generator: PhaseSpaceGenerator,
```

Comment:
May have to think of some structure to provide default arguments, same for `random_generator` for simpler scripts (as discussed). In that case, would also have to change the order of the arguments:
```python
def generate_data(
    size: int,
    intensity: Function,
    kinematics: Kinematics,
    phsp_generator: <?str or Enum?>,
    random_generator: <?UniformRealNumberGenerator, not sure how to provide default argument here>,
) -> np.ndarray:
```

Reply:
The current code addresses this issue. However we might have to think about this a bit more, as the dependency injection we use currently assumes that all constructors of phase space generators and random number generators are the same (which does not have to be true)